### PR TITLE
batches: build clone url on demand

### PR DESCRIPTION
**Pull request by asdine**
This replaces usages of Sources.CloneURL in the batches/reconcilier package with building the url with the external service configuration.
This also changes the sources.ChangesetSource implementations, which now take a context and an external service store.

### CODE REVIEW

File: enterprise/internal/batches/sources/sources_test.go [hunk 1]
```diff
@@ -202,109 +225,232 @@ func TestGitserverPushConfig(t *testing.T) {
 	}
 	tcs := []struct {
 		name                string
+		repoName            string
 		externalServiceType string
-		cloneURL            string
+		config              string
+		repoMetadata        interface{}
+		overrideRepoURL     string
```
Comment (eseliger) [inline enterprise/internal/batches/sources/sources_test.go:232]: I think this is unused

**Review by eseliger — APPROVED** [review]
LGTM

**Review by unknwon — APPROVED** [review]

**Review by LawnGnome — APPROVED** [review]
I second @eseliger's review; otherwise this looks good. :+1:

File: enterprise/internal/batches/sources/sources_test.go [hunk 2]
```diff
@@ -10,68 +11,92 @@ import (
...
 		{
-			name:      "ssh no protocol specified",
-			want:      "ssh://git@github.com/sourcegraph/sourcegraph.git",
-			cloneURLs: []string{"git@github.com:sourcegraph/sourcegraph.git"},
+			name: "https user password",
+			want: "https://git:secrettoken@github.com/sourcegraph/sourcegraph",
+			configs: []string{
+				`{"url": "https://github.com", "authorization": {}}`,
+			},
+			overrideRepoURL: "https://git:secrettoken@github.com/sourcegraph/sourcegraph",
```
[the remark this answers is not in the document]
Comment (asdine) [reply enterprise/internal/batches/sources/sources_test.go:41]: I don't think so, but since the test was already there I adapted it. I'll remove it 👍🏼

**Pull request merged**